### PR TITLE
Add is_tensorlist_like_type helper

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* #123129
* #123110
* __->__ #123109
* #123108

Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

### CODE REVIEW

File: test/test_custom_ops.py [hunk 1]
```diff
@@ -1730,6 +1730,28 @@ def test_is_functional_schema(self):
...
 
+    def test_is_tensorlist_like_type(self):
+        tensorlists = [
+            # Tensor[]
+            torch.ops.aten.where.default._schema.returns[0].type,
+            # Tensor?[]
+            torch.ops.aten.index.Tensor._schema.arguments[1].type,
+            # Tensor[]?
+            torch._C.parse_schema("foo(Tensor[]? x) -> ()").arguments[0].type,
+            # Tensor?[]?
```
Comment (albanD) [inline test/test_custom_ops.py:1741]: Are these actually valid schemas??

File: torch/_library/utils.py [hunk 2]
```diff
@@ -103,6 +104,15 @@ def is_functional(schema):
     return is_functional(schema)
 
 
+def is_tensorlist_like_type(typ: torch.Type):
```
Comment (albanD) [inline torch/_library/utils.py:107]: I guess we have this in torchgen already? Do we expect deduplication to be able to happen between the two or they can't depend on each other?

File: test/test_custom_ops.py (hunk 1 above)
Comment (zou3519) [reply test/test_custom_ops.py:1741]: It's unclear to me?

File: torch/_library/utils.py (hunk 2 above)
Comment (albanD) [reply torch/_library/utils.py:107]: Ho? Which ones don't?

File: test/test_custom_ops.py (hunk 1 above)
Comment (albanD) [reply test/test_custom_ops.py:1741]: I'm surprised that parse_schema doesn't raise....

Comment (zou3519) [reply test/test_custom_ops.py:1741]: Yes, these are all valid (and used)

File: torch/_library/utils.py (hunk 2 above)
Comment (zou3519) [reply torch/_library/utils.py:107]: resolved offline: long-term state is that we want deduplication but some internal builds make it difficult to deduplicate.

**Review by albanD — APPROVED** [review]
Ok!

**Pull request closed without merging**